### DISCUSSION TITLE
用自己数据集训练的识别模型的train_model转换为inference_model后识别结果错误

### DISCUSSION THREAD

**Discussion by zhoushijie28** [PaddlePaddle/PaddleOCR#14088] [Q&A] [answered]
我用自己的数据集训练了识别模型，用训练模型测试时检测结果全对，但是经过export_model.py转换为推理模型后识别结果就几乎都不对了，结果如下：
这是原图
![IMG_20240903_083113_crop_2](https://github.com/user-attachments/assets/c62c9d9a-cb6e-41ae-82dd-c1598631b4f8)
这是训练模型识别结果
![image](https://github.com/user-attachments/assets/87ca907f-49b9-4654-992c-d8994752e2f1)
这是推理模型识别结果
![image](https://github.com/user-attachments/assets/62e9da65-491e-4ffb-b909-d5714f6e0e30)
我的export_model.py的形参是-c "../configs/rec/en_new_finetune_rec.yml" -o Global.pretrained_model=../output/en_finetune_rec/best_accuracy.pdparams Global.save_inference_dir="../rec/inference_model/rec/"
离谱的是之前训练和转换的结果都是对的，代码也都没改，前两天重新训练后就变成了这样，求教一下各位大佬这可能是因为啥造成的啊

**Comment by amazoncer** [comment]
我也遇到同样的问题请问你的解决了吗

**Comment by zhoushijie28** [comment, marked as answer]
解决了，我那个错误是因为用英文字典训练的，推理的时候没有指定字典，所以按照默认的中英文字典进行推理，识别的结果就不对。
在推理的时候指定正确的字典路径就可以了，应该有一个参数是指定字典路径的



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年12月19日(周四) 中午11:26
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [PaddlePaddle/PaddleOCR] 用自己数据集训练的识别模型的train_model转换为inference_model后识别结果错误 (Discussion #14088)




 
我也遇到同样的问题请问你的解决了吗

**Reply by amazoncer** [reply]
你好我也是使用预训练模型进行训练的，但是训练后转推理模型，预测简单的电脑文字截图准确率几乎为0，虽然训练使用的图集并不多，但是预训练模型是官方预训练过的，有一定的准确率才对。你知道这种是啥原因吗。

**Comment by SWHL** [comment]
训练时，验证集指标是多少呀

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年12月21日 19:00 |
| 收件人 | PaddlePaddle/PaddleOCR ***@***.***> |
| 抄送人 | SWHL ***@***.***>,
State ***@***.***> |
| 主题 | Re: [PaddlePaddle/PaddleOCR] 用自己数据集训练的识别模型的train_model转换为inference_model后识别结果错误 (Discussion #14088) |

你好我也是使用预训练模型进行训练的，但是训练后转推理模型，预测简单的电脑文字截图准确率几乎为0，虽然训练使用的图集并不多，但是预训练模型是官方预训练过的，有一定的准确率才对。你知道这种是啥原因吗。

**Comment by zhoushijie28** [comment]
我当时自己训练的指标还是挺高的，训练都没有问题，你的训练集是多大的呢，有没有试过直接用官方的预训练模型测试一下看看效果，可以加一下我的QQ交流2206120168



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年12月21日(周六) 晚上8:01
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [PaddlePaddle/PaddleOCR] 用自己数据集训练的识别模型的train_model转换为inference_model后识别结果错误 (Discussion #14088)




 训练时，验证集指标是多少呀
 
 ---- 回复的原邮件 ----
 | 发件人 | ***@***.***&gt; |
 | 发送日期 | 2024年12月21日 19:00 |
 | 收件人 | PaddlePaddle/PaddleOCR ***@***.***&gt; |
 | 抄送人 | SWHL ***@***.***&gt;,
 State ***@***.***&gt; |
 | 主题 | Re: [PaddlePaddle/PaddleOCR] 用自己数据集训练的识别模型的train_model转换为inference_model后识别结果错误 (Discussion #14088) |
 
 你好我也是使用预训练模型进行训练的，但是训练后转推理模型，预测简单的电脑文字截图准确率几乎为0，虽然训练使用的图集并不多，但是预训练模型是官方预训练过的，有一定的准确率才对。你知道这种是啥原因吗。

**Reply by SWHL** [reply]
识别模型一般看字典大小。如果中英文都有的话，官方给的模型用了1700w数据训练。在技术报告中说了。
如果是英文+字符的话，可能小一些。可以先在discussion中搜一下，有很多相关训练的帖子。

**Comment by karanpate1** [comment]
> I trained a recognition model using my own dataset. When testing with the trained model, the detection results were all correct. However, after converting it to an inference model using export_model.py, the recognition results are almost all incorrect. The results are as follows: This is the original image This is the recognition result of the trained model This is the recognition result of the inference model The parameters in my export_model.py are -c "../configs/rec/en_new_finetune_rec.yml" -o Global.pretrained_model=../output/en_finetune_rec/best_accuracy.pdparams Global.save_inference_dir="../rec/inference_model/rec/" What's weird is that the training and conversion results were correct before, and I didn't change the code. But after retraining a few days ago, the results became like this. Could anyone please tell me what might be causing this? ![IMG_20240903_083113_crop_2](https://private-user-images.githubusercontent.com/85613876/379987878-c62c9d9a-cb6e-41ae-82dd-c1598631b4f8.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3OTQwMzIsIm5iZiI6MTc1ODc5MzczMiwicGF0aCI6Ii84NTYxMzg3Ni8zNzk5ODc4NzgtYzYyYzlkOWEtY2I2ZS00MWFlLTgyZGQtYzE1OTg2MzFiNGY4LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI1VDA5NDg1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyY2M3NTYzMGIyNjRiZTQyYzc2MjBkNTllYjg1MThlOTExOWY4ZWVlNDEyZjdiN2YyY2NhNjY5MjI2ZWRhZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MAv2E8URESOIwKPT08O57p-uI1dQkNqyDcuO2j_WYDA) ![image](https://private-user-images.githubusercontent.com/85613876/379988027-87ca907f-49b9-4654-992c-d8994752e2f1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3OTQwMzIsIm5iZiI6MTc1ODc5MzczMiwicGF0aCI6Ii84NTYxMzg3Ni8zNzk5ODgwMjctODdjYTkwN2YtNDliOS00NjU0LTk5MmMtZDg5OTQ3NTJlMmYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI1VDA5NDg1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzOGUyZTczNmYzMzVhZTE4NGNhODY5MDVlYmRlYzZjNTZiMzU0ZTNhMzZjZTUzNjllMDBlNzVlYWY0NDU3ZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jvEwjjSFRakCtAbUxSYJQFsK29OIS4xCD54gJYoA1b0) ![image](https://private-user-images.githubusercontent.com/85613876/379988164-62e9da65-491e-4ffb-b909-d5714f6e0e30.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3OTQwMzIsIm5iZiI6MTc1ODc5MzczMiwicGF0aCI6Ii84NTYxMzg3Ni8zNzk5ODgxNjQtNjJlOWRhNjUtNDkxZS00ZmZiLWI5MDktZDU3MTRmNmUwZTMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI1VDA5NDg1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2M2I0OWRhYjVkMjYwZDdiYzlhZjcwM2NmMGYwYjlhMDJjNmFlYWE5OWE4ZGVlZjhkYjUyNDEwMDQ0NDM3ZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.e2hPEyWhzCxau8Pm7E7tIwijpr1leBS8mddRuqsVWyk)


###  Issue: High Validation Accuracy but Garbage Inference in PP-OCRv5 (Solved)

I fine-tuned **PP-OCRv5** on a custom dataset (English alphabets, numerals, and some symbols).

* Training accuracy: **98%**
* Validation accuracy: **99%**

But when I ran inference (even on images from training/validation sets), the predictions were totally wrong.

I double-checked everything:

* Followed official finetuning + export steps
* Tested with both `tools/infer_rec.py` and exported inference model
* Same problem: **terrible results**

At this point, I was stuck because evaluation showed **99% accuracy**, but inference was useless.

---

###  Root Cause

Turns out the problem was due to **mismatch in image preprocessing between training and inference**.

In my training config `PP-OCRv5_mobile_rec.yml`:

```yaml
sampler:
  name: MultiScaleSampler
  scales: [[320, 32], [320, 48], [320, 64]]
```

So during training, images were dynamically resized to **3 different shapes**:

* 3×32×320
* 3×48×320
* 3×64×320

But in my exported `inference.yml`, the preprocessing looked like this:

```yaml
PreProcess:
  transform_ops:
  - DecodeImage:
      channel_first: false
      img_mode: BGR
  - MultiLabelEncode:
      gtc_encode: NRTRLabelEncode
  - RecResizeImg:
      image_shape:
      - 3
      - 48
      - 320
```

Notice the mismatch?

* Training: random **multi-scale** (32, 48, 64 heights)
* Inference: fixed to **48 height only**

That mismatch caused the network to completely fail at inference.

---

### ✅ Fix

I manually adjusted the `inference.yml` to try different scales from training.

* With **3×48×320**, results were trash.
* With **3×64×320**, also trash.
* With **3×32×320**, predictions became **perfect**, matching my training/validation accuracy.

So the fix was:
 **Set `RecResizeImg.image_shape` in `inference.yml` to the same shape(s) used in training.**

Final fix:

```yaml
PreProcess:
  transform_ops:
  - DecodeImage:
      channel_first: false
      img_mode: BGR
  - MultiLabelEncode:
      gtc_encode: NRTRLabelEncode
  - RecResizeImg:
      image_shape:
      - 3
      - 32
      - 320
```
---

###  Takeaway

If you’re fine-tuning PP-OCR models with `MultiScaleSampler`, make sure your **inference preprocessing matches your training scales**. Otherwise, you’ll see high validation accuracy but garbage results at inference.

---

This should save people from wasting days like I did. 🙃

---

Want me to polish this into a proper **GitHub issue comment** format (with Markdown headings + code blocks, short and punchy)? Or do you want it more like a **blog-style post** with more storytelling?
